### DISCUSSION TITLE
Do tasks inherit tracing spans?

### DISCUSSION THREAD

**Discussion by BGR360** [tokio-rs/tokio#6008] [Q&A] [answered]
When a task is spawned, does it inherit the current tracing span? I don't see this called out in the docs, nor am I able to easily figure it out by reading the tokio source code. I (think that I) want this behavior, because I want to be able to associate spawned tasks with the original operations with which they were meant to help.

**Comment by hawkw** [comment, marked as answer]
Tasks don't inherit the current `tracing` span unless the current span is explicitly propagated to the spawned task. For example, if you want to spawn a task in the same span as the context in which it's spawned, you can use [`Instrument::in_current_span`](https://docs.rs/tracing/latest/tracing/trait.Instrument.html#method.in_current_span):
```rust
use tracing::Instrument;

tokio::spawn(my_task().in_current_span());
// which is equivalent to
tokio::spawn(my_task().instrument(tracing::Span::current());
```

Alternatively, if you instrument the spawned future with a new span, that span will have the current span as its _parent_, so that task will be conceptually "inside" the current span:
```rust
use tracing::{Instrument, info_span};

// creates a new span called `my_task`, which is a child of the current span, and
// attaches it to the spawned future.
tokio::spawn(my_task().instrument(info_span!("my_task"));
```

Finally, if you want a spawned task to inherit the current task's span if the child span is not enabled by filtering, you can use [`Span::or_current`](https://docs.rs/tracing/latest/tracing/struct.Span.html#method.or_current):
```rust
use tracing::{Instrument, info_span};

// creates a new span called `my_task`, which is a child of the current span:
let span = info_span!("my_task")
     // if the child span is not enabled by the `tracing` subscriber, use the 
     // current span, instead:
     .or_current();
tokio::spawn(my_task().instrument(span));

**Reply by BGR360** [reply]
Heck yes, that is so helpful and cool. Thank you!

**Reply by BGR360** [reply]
If the future I'm spawning is the output of an `#[instrument]`-ed `async fn`, will it have the current span as its parent without me calling `.instrument()` on it? I thought I remembered seeing something about `#[instrument]` spitting out `Instrumented` futures already.

**Reply by hawkw** [reply]
Yes, the span generated by `#[instrument]` will have the current span when the `async fn` is called as its parent. Putting `#[instrument]` on an `async fn` is essentially saying "every time I call this `async fn`, I want to immediately `.instrument()` it with this span".

**Comment by tombh** [comment]
I'd just like to add one more bit of advice to @hawkw's excellent answer. When using the `async move {}` pattern, the current span isn't actually automatically included in the `tokio::spawn()`. So a pattern like this is needed:
```rust
let current_span = tracing::Span::current();
tokio::spawn(async move { my_task().instrument(current_span) })
```
